library/src/main/java/com/novoda/downloadmanager/VersionOneToVersionTwoMigrator.java

```diff
@@ -0,0 +1,143 @@
+package com.novoda.downloadmanager;
+
+import android.util.Log;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.util.List;
+
+import static com.novoda.downloadmanager.DownloadBatchStatus.Status;
+
+class VersionOneToVersionTwoMigrator implements Migrator {
+
+    private static final int BUFFER_SIZE = 8 * 512;
+    private static final String DELETE_QUERY = "DELETE FROM batches WHERE _id = ?";
+
+    private final MigrationExtractor migrationExtractor;
+    private final RoomDownloadsPersistence downloadsPersistence;
+    private final InternalFilePersistence internalFilePersistence;
+    private final SqlDatabaseWrapper database;
+    private final Callback migrationCompleteCallback;
+
+    VersionOneToVersionTwoMigrator(MigrationExtractor migrationExtractor,
+                                   RoomDownloadsPersistence downloadsPersistence,
+                                   InternalFilePersistence internalFilePersistence,
+                                   SqlDatabaseWrapper database,
+                                   Callback migrationCompleteCallback) {
+        this.migrationExtractor = migrationExtractor;
+        this.downloadsPersistence = downloadsPersistence;
+        this.internalFilePersistence = internalFilePersistence;
+        this.database = database;
+        this.migrationCompleteCallback = migrationCompleteCallback;
+    }
+
+    @Override
+    public void migrate() {
+        List<Migration> migrations = migrationExtractor.extractMigrations();
+
+        migrateV1FilesToV2Location(migrations);
+        migrateV1DataToV2Database(migrations);
+
+        deleteFrom(database, migrations);
+        database.close();
+
+        migrationCompleteCallback.onMigrationComplete();
+    }
+
+    private void migrateV1FilesToV2Location(List<Migration> migrations) {
+        for (Migration migration : migrations) {
+            Batch batch = migration.batch();
+            List<FileSize> fileSizes = migration.fileSizes();
+            List<String> originalFileLocations = migration.originalFileLocations();
+
+            migrateV1FilesToV2Location(batch, fileSizes, originalFileLocations);
+        }
+    }
+
+    private void migrateV1FilesToV2Location(Batch batch, List<FileSize> fileSizes, List<String> originalFileLocations) {
+        for (int i = 0; i < originalFileLocations.size(); i++) {
+            String originalFileLocation = originalFileLocations.get(i);
+            FileSize actualFileSize = fileSizes.get(i);
+            FileName newFileName = LiteFileName.from(batch, batch.getFileUrls().get(i));
+            internalFilePersistence.create(newFileName, actualFileSize);
+
+            FileInputStream inputStream = null;
+            try {
+                // open the v1 file
+                inputStream = new FileInputStream(new File(originalFileLocation));
+                byte[] bytes = new byte[BUFFER_SIZE];
+
+                // read the v1 file
```

Comment:
I wonder if we could extract a general file copier from here? Might reduce length + nesting, tighten scope. The comments seem to back up that something should be extracted.

I do notice that it writes to that internalFilePersistence thing rather than an output stream, which would make that not so simple. Though that might be a prompt to make the API of that more in keeping with typical Java File and Stream IO, return a stream when creating a file or something.